### PR TITLE
Simple heatmap implementation example

### LINKED ISSUE #603: Simple heatmap implementation
When the parameter `&view=CH.SIA380-1.heatmap` is supplied in the URL, the following change is applied to the rendering of the model.

- Across all elements of the model find the MIN and the MAX values of the attributes PSet_vyzn.Verlust or PSet_Vyzn.SIA380-1.TransmissionHeatLoss 
- Color **each** element according to these rules
  - 0 or undefined -> #F0F0F0
  - Positive values ->  Scale colors linearly from #F0F0F0 (min) to #EB3324 (max)
  - Negative values -> Scale colors linearly from #F0F0F0 (min) to #22B14C (max)

See attached model for an example.
[220714_seestrasse Heatmap (2).zip](https://github.com/bldrs-ai/Share/files/10744693/220714_seestrasse.Heatmap.2.zip)

### PR DESCRIPTION
**Pull request by aozien**
**This PR closes #603** 

**TL;DR** add `&view=CH.SIA380-1.heatmap`  to the url, it will disappear but the setting will be applied nontheless

~~It shouldn't be merged with main, it's just a quick demo of~~ 
visualizing a property value as a heat map color scale from #110000 to #FF0000

Adding query parameters will be cleared upon model reload, so as a work around, the main container cadview 'remembers' whether or not the parameter `&view=CH.SIA380-1.heatmap`  existed in the url, and will use the view rules in all subsequent model loads.

### CODE REVIEW

**Comment by pablo-mayrgundter** [comment]
I'm unclear on the status.. could you update the PR description with status and provide a demo link?

**Comment by aozien** [comment]
Sorry, I missed this reply, This PR is ready to be merged.

This special coloring mode can be accessed by adding the view parameter to the url, for example

https://deploy-preview-605--bldrs-share.netlify.app/share/v/p/index.ifc?view=CH.SIA380-1.heatmap

File: src/Infrastructure/ColorHelperFunctions.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import IfcColor from './IfcColor'
+
+/**
+ * Convert Hex Color string to IfcColor Object
+ *
+ * @param {string} hexColor the color in string format '#C7C7C7'
+ * @return {IfcColor} the IfcColor object
+ */
+export function parseColor(hexColor) {
```
Comment (pablo-mayrgundter) [inline src/Infrastructure/ColorHelperFunctions.js:9]: Please add unit tests

File: src/Infrastructure/ColorHelperFunctions.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * get a color that lies on the scale between two colors
+ *
+ * @param {IfcColor} startColor
+ * @param {IfcColor} targetColor
+ * @param {number} value
+ * @param {number} min value
+ * @param {number} max value
+ * @return {IfcColor} color on the scale between the start color and max color
+ */
+export function interpolateColors(startColor, targetColor, value, min, max) {
```
Comment (pablo-mayrgundter) [inline src/Infrastructure/ColorHelperFunctions.js:29]: Please add unit tests

File: src/Infrastructure/ColorHelperFunctions.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * convert value from one value scale to another
+ *
+ * @param {number} value
+ * @param {number} min
+ * @param {number} max
+ * @param {number} targetMin
+ * @param {number} targetMax
+ * @return {number} the new value
+ */
+export function changeValueScale(value, min, max, targetMin, targetMax) {
```
Comment (pablo-mayrgundter) [inline src/Infrastructure/ColorHelperFunctions.js:48]: Please add unit tests

File: src/Infrastructure/ViewRulesCompiler.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+/* eslint-disable no-magic-numbers */
+/**
+ * Convert view rules into view settings object
+ *
+ * @param {object} api
+ * @param {number} modelID
+ * @param {number} rules
+ * @return {IfcCustomViewSettings} the view settings
+ */
+export async function compileViewRules(api, modelID, rules) {
```
Comment (pablo-mayrgundter) [inline src/Infrastructure/ViewRulesCompiler.js:17]: Please add unit tests

File: src/Containers/CadView.jsx [hunk 5]
```diff
@@ -33,6 +33,7 @@ import {addSceneLayer} from './SceneLayer'
 export const searchIndex = new SearchIndex()
 let count = 0
 
+const useViewRules = window.location.href.match(/(view=CH.SIA380-1.heatmap)/ig)?.length > 0
```
Comment (pablo-mayrgundter) [inline src/Containers/CadView.jsx:36]: Can you move this to IfcViewerApiExtended?  We're trying to limit any changes to CadView, so should avoid any changes and looks like this doesn't actually have a dep.

File: src/Infrastructure/ViewRulesCompiler.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  const allPropObjects = []
+  for (let i = 0; i < propRelLines.size(); i++) {
+    allPropObjects.push( await api.GetLine(modelID, propRelLines.get(i), true))
+  }
+  const psetObjects = allPropObjects.filter((x) => x.RelatingPropertyDefinition.type === IFCPROPERTYSET)
+  // Get only sets containing PSet_vyzn.Verlust
+  const objectsAndPropVal = psetObjects.map((a) =>
+    ({o: a.RelatedObjects[0]?.expressID,
+      p: a.RelatingPropertyDefinition.HasProperties?.find(
+          (s) => s.Name.value === 'Verlust' ||
```
Comment (pablo-mayrgundter) [inline src/Infrastructure/ViewRulesCompiler.js:29]: I wonder if these ought to be passed by URL param and have the application be stateless..

- on load, look for StateToken encoding param names to be colorized, and a color map
- recolor mesh in scene (unload and reload mesh if need be?)
- no state kept

**Review by pablo-mayrgundter — COMMENTED** [review]
Hey Zien, I think this is great functionality, but still a little ambivalent about the kind of demo status of this.  I have a proposal inline about passing more params to generalize this.  See what you think

File: src/Infrastructure/ViewRulesCompiler.js (hunk 6 above)
Comment (aozien) [reply src/Infrastructure/ViewRulesCompiler.js:29]: Hi @pablo-mayrgundter The current 'ViewRulesCompiler' file is a just dummy implementation that mimics the compilation of a bunch of rules against the IFC model on loading, and returns a result of 'IfcCustomViewSettings' to be used for overriding model colors.

The compiler functionality itself, and how the 'ViewRules' are stored and shared or changed will be explored in details in #475, that way it would be more generic and flexible, but for now it's fixed with the predefined view rules in #603

They're both big topics (compiling and storing/changing the ViewRules) that this PR will get too complicated if it tried to cover both of them as well

this PR's main focus was validating the workflow of having a predefined set of rules, that would be compiled and applied on model load, that's why there's not much focus on the actual compilation of viewRules itself

File: src/Infrastructure/ViewRulesCompiler.js (hunk 4 above)
Comment (aozien) [reply src/Infrastructure/ViewRulesCompiler.js:17]: This is currently just a placeholder for the rule compilation functionality
